### ISSUE TITLE
Bug: std::filesystem::copy_file("/zip/file.txt", "extracted-file.txt") fails

### ISSUE THREAD

**Issue by jcbhmr** [jart/cosmopolitan#1248]
### Contact Details

jcbhmr@outlook.com

### What happened?

This program:

```cpp
#include <filesystem>

int main() {
    std::filesystem::copy_file("/zip/file.txt", "./file-extracted.txt");
    return 0;
}
```

```
Hello from file.txt!
```

```sh
cosmocc -o ./app.com ./app.cpp
zip -A ./app.com ./file.txt
```

...should copy the embedded file to `./file-extracted.txt`. It fails. I don't know why.

```
libc++abi: terminating due to uncaught exception of type std::__1::__fs::filesystem::filesystem_error: filesystem error: in copy_file: Invalid argument ["./file-extracted.txt"] ["/zip/file.txt"]
Aborted
```

The same program when compiled with my system C++ compiler (WSL Linux Ubuntu 22.04 x64) works fine when I change `/zip/file.txt` to `./file.txt`.

The same program works when I use a non-`/zip/*` path with `cosmocc`. That makes me think it's something to do with `/zip/*` that `std::filesystem::copy_file()` doesn't like.

```cpp
#include <filesystem>

int main() {
    std::filesystem::copy_file("./file.txt", "./file-extracted.txt");
    return 0;
}
```

👆 this works. `/zip/file.txt` doesn't.

My current workaround is to re-implement my own `filesystemzipfix::copy_file()` function that uses code like this:

```cpp
auto ifs = std::ifstream(from, std::ios::binary);
auto ofs = std::ofstream(to, std::ios::binary);
ofs << ifs.rdbuf();
ifs.close();
ofs.close();
std::filesystem::permissions(std::filesystem::status(from).permissions());
```

-ish. I don't know if that's the right way to do things but it appeared to work well enough.

The reason I want `std::filesystem::copy_file()` to work is that I want to use `std::filesystem::copy()` to recursively copy a `/zip/platform-specific-files-for-windows-like-other-exes/*` and `/zip/platform-specific-files-for-linux-that-need-to-be-real-files/*` bundle to `%LOCALAPPDATA%/my-app/cache` or similar. Right now `std::filesystem::copy_file()` and `std::filesystem::copy()` don't seem to work.

OH I SHOULD ALSO NOTE that when it fails it _creates the file_ it just doesn't have any content. It's like the copy operation failed but the create-the-file operation succeeded. 🤷‍♀️

### Version

cosmocc (GCC) 14.1.0

### What operating system are you seeing the problem on?

Linux

### Relevant log output

```shell
SYS 808937 808937             67'499 mmap(0x6fe000000, 4'096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED, -1, 0) → 0x6fe000000 (8'892'416 bytes total)
SYS 808937 808937            109'504 getcwd([u"/home/jcbhmr/Documents/test12345"], 1'023) → 33
SYS 808937 808937            119'110 getenv("COSMOPOLITAN_DISABLE_ZIPOS") → NULL
SYS 808937 808937            121'673 getenv("COSMOPOLITAN_INIT_ZIPOS") → NULL
SYS 808937 808937            124'023 issetugid() → 0
SYS 808937 808937            126'026 GetProgramExecutableName() → "/home/jcbhmr/Documents/test12345/app.com"
SYS 808937 808937            134'570 openat(AT_FDCWD, "/home/jcbhmr/Documents/test12345/app.com", O_RDONLY) → 3
SYS 808937 808937            163'475 fstat(3, [{.st_size=1'035'950, .st_blocks=1'036'288/512, .st_mode=0100755, .st_uid=1000, .st_gid=1000, .st_dev=0x820, .st_ino=0x409e7}]) → 0
SYS 808937 808937            195'243 mmap(0, 1'035'950, PROT_READ, MAP_SHARED, 3, 0) → 0x7f1cab6df000 (9'928'704 bytes total)
SYS 808937 808937            242'485 munmap(0x7f1cab6df000, 970'752) → 0 (8'957'952 bytes total)
SYS 808937 808937            278'658 madvise(0x7f1cab7cc000, 61'440, 4) → 0
SYS 808937 808937            307'135 mmap(0, 32, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x7f1cab7cb000 (8'962'048 bytes total)
SYS 808937 808937            338'331 close(3) → 0
SYS 808937 808937            360'985 __zipos_get("/home/jcbhmr/Documents/test12345/app.com") → 0
SYS 808937 808937            391'583 mmap(0, 139'264, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x7f1cab7a9000 (9'101'312 bytes total)
SYS 808937 808937          1'571'309 inflate([u"SYMT☺   ≡○        ☻       ☻       @     "...], 139'264, u"∞▌gxT╒┬0∞5t♦%b☺+▒ ╪ ☻◙VBæÇ¿♦DA╘ôPBB═↑☻AA"..., 34'835) → 0
SYS 808937 808937          1'615'206 GetSymbolTableFromZip() → 0x7f1cab7a9000
SYS 808937 808937          1'618'955 getenv("TERM") → "xterm-256color"
SYS 808937 808937          1'623'102 sys_getrandom(0,0,0) → 0
SYS 808937 808937          1'650'875 getenv("COSMOPOLITAN_HEAP_COUNT") → NULL
SYS 808937 808937          1'680'041 mmap(0, 65'536, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x7f1cab799000 (9'166'848 bytes total)
SYS 808937 808937          1'713'478 mmap(0, 65'536, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x7f1cab789000 (9'232'384 bytes total)
SYS 808937 808937          1'742'828 created 2 dlmalloc arenas for 4 cpus
SYS 808937 808937          1'778'938 mmap(0, 64, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x7f1cab788000 (9'236'480 bytes total)
SYS 808937 808937          1'810'546 openat(AT_FDCWD, "/zip/file.txt", O_RDONLY|O_NONBLOCK) → 3
SYS 808937 808937          1'835'160 fstat(3, [{.st_size=6, .st_blocks=512/512, .st_mode=0100644, .st_dev=0x409e7, .st_ino=0xfce4a}]) → 0
SYS 808937 808937          1'843'486 fstatat(AT_FDCWD, "./file-extracted.txt", [n/a], 0) → -1 ENOENT
SYS 808937 808937          1'930'208 openat(AT_FDCWD, "./file-extracted.txt", O_WRONLY|O_CREAT, 0100644) → 4 ENOENT
SYS 808937 808937          1'957'154 fstat(4, [{.st_size=0, .st_mode=0100644, .st_uid=1000, .st_gid=1000, .st_dev=0x820, .st_ino=0x3dcdf}]) → 0 ENOENT
SYS 808937 808937          1'962'895 sendfile(4, 3, 0x45f30e, 6) → -1 EINVAL
SYS 808937 808937          1'980'528 getenv("LIBUNWIND_PRINT_APIS") → NULL
SYS 808937 808937          1'983'590 getenv("LIBUNWIND_PRINT_UNWINDING") → NULL
libc++abi: SYS 808937 808937          2'001'090 writev(2, {{u"", 0}, {u"libc++abi: ", 11}}, 2) → 11 EINVAL
terminating due to uncaught exception of type std::__1::__fs::filesystem::filesystem_error: filesystem error: in copy_file: Invalid argument ["./file-extracted.txt"] ["/zip/file.txt"]SYS 808937 808937          2'008'331 writev(2, {{u"", 0}, {u"terminating due to uncaught exception of"..., 183}}, 2) → 183 EINVAL

SYS 808937 808937          2'013'029 writev(2, {{u"", 0}, {u"◙", 1}}, 2) → 1 EINVAL
SYS 808937 808937          2'017'370 sigprocmask(SIG_UNBLOCK, {ABRT}, [NULL]) → 0 EINVAL
Aborted
```

**Comment by jcbhmr** [comment]
For reference, here's the --strace of the working `cosmoc++ -o ./app.com ./app.cpp` that uses `std::filesystem::copy_file("./file.txt", "./file-extracted.txt")`

```
SYS 810402 810402             82'702 mmap(0x6fe000000, 4'096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED, -1, 0) → 0x6fe000000 (8'892'416 bytes total)
SYS 810402 810402            164'369 getcwd([u"/home/jcbhmr/Documents/test12345"], 1'023) → 33
SYS 810402 810402            197'213 getenv("COSMOPOLITAN_DISABLE_ZIPOS") → NULL
SYS 810402 810402            203'061 getenv("COSMOPOLITAN_INIT_ZIPOS") → NULL
SYS 810402 810402            205'621 issetugid() → 0
SYS 810402 810402            209'290 GetProgramExecutableName() → "/home/jcbhmr/Documents/test12345/app.com"
SYS 810402 810402            233'819 openat(AT_FDCWD, "/home/jcbhmr/Documents/test12345/app.com", O_RDONLY) → 3
SYS 810402 810402            240'730 fstat(3, [{.st_size=1'035'799, .st_blocks=970'752/512, .st_mode=0100755, .st_uid=1000, .st_gid=1000, .st_dev=0x820, .st_ino=0x409e7}]) → 0
SYS 810402 810402            247'266 mmap(0, 1'035'799, PROT_READ, MAP_SHARED, 3, 0) → 0x7f2c87846000 (9'928'704 bytes total)
SYS 810402 810402            289'619 munmap(0x7f2c87846000, 970'752) → 0 (8'957'952 bytes total)
SYS 810402 810402            340'877 madvise(0x7f2c87933000, 61'440, 4) → 0
SYS 810402 810402            383'443 mmap(0, 24, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x7f2c87932000 (8'962'048 bytes total)
SYS 810402 810402            413'976 close(3) → 0
SYS 810402 810402            437'024 __zipos_get("/home/jcbhmr/Documents/test12345/app.com") → 0
SYS 810402 810402            467'710 mmap(0, 139'264, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x7f2c87910000 (9'101'312 bytes total)
SYS 810402 810402          1'694'671 inflate([u"SYMT☺   ≡○        ☻       ☻       @     "...], 139'264, u"∞▌gxT╒┬0∞5t♦%b☺←─◙6êÇéòP$ *☺Q►⌡$öÉP3å@PP"..., 34'834) → 0
SYS 810402 810402          1'721'099 GetSymbolTableFromZip() → 0x7f2c87910000
SYS 810402 810402          1'724'732 getenv("TERM") → "xterm-256color"
SYS 810402 810402          1'748'037 sys_getrandom(0,0,0) → 0
SYS 810402 810402          1'781'333 getenv("COSMOPOLITAN_HEAP_COUNT") → NULL
SYS 810402 810402          1'809'553 mmap(0, 65'536, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x7f2c87900000 (9'166'848 bytes total)
SYS 810402 810402          1'842'351 mmap(0, 65'536, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x7f2c878f0000 (9'232'384 bytes total)
SYS 810402 810402          1'885'603 created 2 dlmalloc arenas for 4 cpus
SYS 810402 810402          1'922'745 openat(AT_FDCWD, "./file.txt", O_RDONLY|O_NONBLOCK) → 3
SYS 810402 810402          1'950'022 fstat(3, [{.st_size=6, .st_blocks=4'096/512, .st_mode=0100644, .st_uid=1000, .st_gid=1000, .st_dev=0x820, .st_ino=0x379b6}]) → 0
SYS 810402 810402          1'982'850 fstatat(AT_FDCWD, "./file-extracted.txt", [n/a], 0) → -1 ENOENT
SYS 810402 810402          2'078'103 openat(AT_FDCWD, "./file-extracted.txt", O_WRONLY|O_CREAT, 0100644) → 4 ENOENT
SYS 810402 810402          2'104'608 fstat(4, [{.st_size=0, .st_mode=0100644, .st_uid=1000, .st_gid=1000, .st_dev=0x820, .st_ino=0x3dcdf}]) → 0 ENOENT
SYS 810402 810402          2'146'339 sendfile(4, 3, 0x45f30e, 6) → 6 ENOENT
SYS 810402 810402          2'172'349 close(4) → 0 ENOENT
SYS 810402 810402          2'176'937 close(3) → 0 ENOENT
SYS 810402 810402          2'198'597 exit(0)
SYS 810402 810402          2'201'827 __cxa_finalize(std::__1::generic_category()::AvoidDestroyingGenericCategory::~AvoidDestroyingGenericCategory(), 0x471348)
SYS 810402 810402          2'211'820 __cxa_finalize(std::__1::system_category()::AvoidDestroyingSystemCategory::~AvoidDestroyingSystemCategory(), 0x471340)
SYS 810402 810402          2'216'865 __cxa_finalize(__gdtoa_Bclear, 0)
SYS 810402 810402          2'219'946 __cxa_finalize(fflush_unlocked, 0)
SYS 810402 810402          2'222'745 _Exit(0)
```

**Comment by jcbhmr** [comment]
Can confirm that this also happens on Windows 10 x64 using PowerShell

```shell
PS C:\Users\jcbhm\Downloads\cosmotest> ./app.com --strace
SYS   8952   6832          4'329'050 mmap(0x6fe000000, 65'536, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED, -1, 0) → 0x6fe000000 (9'015'296 bytes total)
SYS   8952   6832          9'077'705 getenv("_COSMO_FDS") → NULL
SYS   8952   6832          9'262'173 getenv("COSMOPOLITAN_DISABLE_ZIPOS") → NULL
SYS   8952   6832          9'455'359 getenv("COSMOPOLITAN_INIT_ZIPOS") → NULL
SYS   8952   6832          9'651'280 GetProgramExecutableName() → "/C/Users/jcbhm/Downloads/cosmotest/app.com"
SYS   8952   6832          9'978'137 openat(AT_FDCWD, "/C/Users/jcbhm/Downloads/cosmotest/app.com", O_RDONLY) → 3
SYS   8952   6832         10'326'786 fstat(3, [{.st_size=1'035'950, .st_blocks=1'036'288/512, .st_mode=0100700, .st_uid=27294, .st_gid=27294, .st_dev=0xb0e7e721, .st_ino=0x1700000004c16e, .st_flags=20}]) → 0
SYS   8952   6832         10'626'353 mmap(0, 1'035'950, PROT_READ, MAP_SHARED, 3, 0) → 0x217c140b0000 (10'051'584 bytes total)
SYS   8952   6832         10'852'787 munmap(0x217c140b0000, 917'504) → 0 EINVAL (10'051'584 bytes total)
SYS   8952   6832         11'021'243 madvise(0x217c1419d000, 61'440, 127) → -1 EINVAL
SYS   8952   6832         11'249'091 mmap(0, 32, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x1169a0db0000 EINVAL (10'055'680 bytes total)
SYS   8952   6832         11'505'438 close(3) → 0 EINVAL
SYS   8952   6832         11'658'557 __zipos_get("/C/Users/jcbhm/Downloads/cosmotest/app.com") → 0 EINVAL
SYS   8952   6832         11'895'882 mmap(0, 139'264, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x1169a0dc0000 EINVAL (10'194'944 bytes total)
SYS   8952   6832         13'445'076 inflate([u"SYMT☺   ≡○        ☻       ☻       @     "...], 139'264, u"∞▌gxT╒┬0∞5t¶%b☺+▒ XÇ◘(X○E☻ó↕►♣QOB○○5c◘EA"..., 34'835) → 0
SYS   8952   6832         13'661'535 GetSymbolTableFromZip() → 0x1169a0dc0000
SYS   8952   6832         13'818'833 getenv("TERM") → "xterm-256color"
SYS   8952   6832         13'989'299 getenv("COSMOPOLITAN_HEAP_COUNT") → NULL
SYS   8952   6832         14'181'117 mmap(0, 65'536, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x1c4888ef0000 (10'260'480 bytes total)
SYS   8952   6832         14'383'653 mmap(0, 65'536, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x1c4888f00000 (10'326'016 bytes total)
SYS   8952   6832         14'566'044 created 2 dlmalloc arenas for 4 cpus
SYS   8952   6832         14'754'618 mmap(0, 64, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) → 0x1c4888f10000 (10'330'112 bytes total)
SYS   8952   6832         14'936'501 openat(AT_FDCWD, "/zip/file.txt", O_RDONLY|O_NONBLOCK) → 3
SYS   8952   6832         15'096'682 fstat(3, [{.st_size=6, .st_blocks=512/512, .st_mode=0100644, .st_dev=0x1700000004c16e, .st_ino=0xfce4a}]) → 0
SYS   8952   6832         15'336'563 fstatat(AT_FDCWD, "./file-extracted.txt", [n/a], 0) → -1 ENOENT
SYS   8952   6832         15'902'631 openat(AT_FDCWD, "./file-extracted.txt", O_WRONLY|O_CREAT, 0100644) → 4 ENOENT
SYS   8952   6832         16'064'067 fstat(4, [{.st_size=0, .st_mode=0100600, .st_uid=27294, .st_gid=27294, .st_dev=0xb0e7e721, .st_ino=0x2300000004fe02, .st_flags=20}]) → 0 ENOENT
SYS   8952   6832         16'212'097 sendfile(4, 3, 0x45f30e, 6) → -1 EBADF
SYS   8952   6832         16'352'037 getenv("LIBUNWIND_PRINT_APIS") → NULL
SYS   8952   6832         16'461'385 getenv("LIBUNWIND_PRINT_UNWINDING") → NULL
libc++abi: SYS   8952   6832         16'623'339 writev(2, {{u"", 0}, {u"libc++abi: ", 11}}, 2) → 11 EBADF
terminating due to uncaught exception of type std::__1::__fs::filesystem::filesystem_error: filesystem error: in copy_file: Bad file number (win32 error 5) ["./file-extracted.txt"] ["/zip/file.txt"]SYS   8952   6832         16'834'130 writev(2, {{u"", 0}, {u"terminating due to uncaught exception of"..., 198}}, 2) → 198 EBADF

SYS   8952   6832         17'018'758 writev(2, {{u"", 0}, {u"◙", 1}}, 2) → 1 EBADF
SYS   8952   6832         17'134'433 sigprocmask(SIG_UNBLOCK, {ABRT}, [NULL]) → 0 EBADF
SYS   8952   6832         17'216'830 terminating on SIGABRT due to no handler
PS C:\Users\jcbhm\Downloads\cosmotest>
```

**Comment by jart** [comment]
This should have been fixed in 0d7c272d3f8b7925b0810d99a6bbc7cfab3a346f. Please try building a new toolchain at head using tool/cosmocc/package.sh and let me know if it's still using sendfile(). Thanks!

**Comment by jcbhmr** [comment]
Building from `master` works! 👍
